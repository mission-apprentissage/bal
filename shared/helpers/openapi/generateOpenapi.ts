import type { ResponseConfig, RouteConfig } from "@asteasolutions/zod-to-openapi";
import { OpenApiGeneratorV31, OpenAPIRegistry } from "@asteasolutions/zod-to-openapi";
import { formatParamUrl } from "@fastify/swagger";
import type { oas30 } from "openapi3-ts";
import type { ZodType } from "zod";

import { zRoutes } from "../../index";
import type { IRouteSchema } from "../../routes/common.routes";
import { ZResError } from "../../routes/common.routes";

function generateOpenApiResponseObject(schema: ZodType, description: string | null = null): ResponseConfig {
  return {
    description: description ?? schema._def.openapi?.metadata?.description ?? schema.description ?? "",
    content: {
      "application/json": {
        schema,
      },
    },
  };
}

const commonResponses: { [statusCode: string]: ResponseConfig | { $ref: string } } = {
  400: { $ref: "#/components/responses/BadRequest" },
  401: { $ref: "#/components/responses/Unauthorized" },
  403: { $ref: "#/components/responses/Forbidden" },
  404: { $ref: "#/components/responses/NotFound" },
  429: { $ref: "#/components/responses/TooManyRequests" },
  500: { $ref: "#/components/responses/InternalServerError" },
  503: { $ref: "#/components/responses/ServiceUnavailable" },
};

function registerResponseError(id: string, description: string, registry: OpenAPIRegistry) {
  registry.registerComponent("responses", id, {
    description,
    content: { "application/json": { schema: { $ref: "#/components/schemas/Error" } } },
  });
}

function registerResponsesError(registry: OpenAPIRegistry) {
  registry.register("Error", ZResError);
  registerResponseError("BadRequest", "La requete ne respecte pas les reglès de validations.", registry);
  registerResponseError("Unauthorized", "L'authentification est requise.", registry);
  registerResponseError("Forbidden", "L'utilisateur n'a pas les droits d'effectuer l'opération demandée", registry);
  registerResponseError("NotFound", "Le serveur n'a pas trouvé la ressource demandée.", registry);
  registerResponseError(
    "TooManyRequests",
    "L'utilisateur a émis trop de requêtes dans un laps de temps donné.",
    registry
  );
  registerResponseError(
    "InternalServerError",
    "Le serveur a rencontré une situation qu'il ne sait pas traiter.",
    registry
  );
  registerResponseError("ServiceUnavailable", "Le serveur n'est pas prêt pour traiter la requête", registry);
}

function generateOpenApiResponsesObject<R extends IRouteSchema["response"]>(
  response: R
): { [statusCode: string]: ResponseConfig | { $ref: string } } {
  const codes = Object.keys(response) as Array<keyof R>;

  return codes.reduce((acc, code: keyof R) => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const schema: ZodType = response[code] as any;
    acc[code as string] = generateOpenApiResponseObject(code in acc ? schema.or(ZResError) : schema);

    return acc;
  }, commonResponses);
}

function generateOpenApiRequest(route: IRouteSchema): RouteConfig["request"] {
  const requestParams: RouteConfig["request"] = {};

  if (route.method !== "get" && route.body) {
    requestParams.body = {
      content: {
        "application/json": { schema: route.body },
      },
      required: true,
    };
  }
  if (route.params) {
    requestParams.params = route.params;
  }
  if (route.querystring) {
    requestParams.query = route.querystring;
  }
  if (route.headers) {
    requestParams.headers = route.headers;
  }

  return requestParams;
}

function getSecurityRequirementObject(route: IRouteSchema): oas30.SecurityRequirementObject[] {
  if (route.securityScheme === null) {
    return [];
  }

  if (route.securityScheme.auth !== "api-key") {
    throw new Error("getSecurityRequirementObject: securityScheme not supported");
  }

  return [{ "api-key": [] }];
}

function addOpenApiOperation(
  path: string,
  method: "get" | "put" | "post" | "delete",
  route: IRouteSchema,
  registry: OpenAPIRegistry
) {
  if (!route.openapi) {
    return;
  }

  const config: RouteConfig = {
    ...route.openapi,
    method,
    path: formatParamUrl(path),
    responses: generateOpenApiResponsesObject(route.response),
  };

  const request = generateOpenApiRequest(route);
  if (request) {
    config.request = request;
  }
  const security = getSecurityRequirementObject(route);
  if (security) {
    config.security = security;
  }

  registry.registerPath(config);
}

export function generateOpenApiSchema(version: string, env: string, publicUrl: string) {
  const registry = new OpenAPIRegistry();

  registry.registerComponent("securitySchemes", "api-key", {
    type: "apiKey",
    name: "authorization",
    in: "header",
  });
  registerResponsesError(registry);

  for (const [method, pathRoutes] of Object.entries(zRoutes)) {
    for (const [path, route] of Object.entries(pathRoutes)) {
      addOpenApiOperation(path, method as "get" | "put" | "post" | "delete", route, registry);
    }
  }

  const generator = new OpenApiGeneratorV31(registry.definitions);

  return generator.generateDocument({
    info: {
      title: "API documentation BAL",
      version,
      license: {
        name: "MIT",
      },
    },
    openapi: "3.1.0",
    servers: [
      {
        url: publicUrl,
        description: env,
      },
    ],
  });
}
