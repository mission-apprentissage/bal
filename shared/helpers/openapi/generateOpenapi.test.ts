import { describe, expect, it } from "vitest";

import { generateOpenApiSchema } from "./generateOpenapi";

describe("generateOpenApiSchema", () => {
  it("should generate proper schema", async () => {
    const s = generateOpenApiSchema("V1.0", "Production", "https://bal.apprentissage.beta.gouv.fr");
    expect(s).toMatchSnapshot();
  });
});
