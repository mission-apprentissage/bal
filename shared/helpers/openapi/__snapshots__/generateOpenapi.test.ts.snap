// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateOpenApiSchema > should generate proper schema 1`] = `
{
  "components": {
    "parameters": {},
    "responses": {
      "BadRequest": {
        "content": {
          "application/json": {
            "schema": {
              "$ref": "#/components/schemas/Error",
            },
          },
        },
        "description": "La requete ne respecte pas les reglès de validations.",
      },
      "Forbidden": {
        "content": {
          "application/json": {
            "schema": {
              "$ref": "#/components/schemas/Error",
            },
          },
        },
        "description": "L'utilisateur n'a pas les droits d'effectuer l'opération demandée",
      },
      "InternalServerError": {
        "content": {
          "application/json": {
            "schema": {
              "$ref": "#/components/schemas/Error",
            },
          },
        },
        "description": "Le serveur a rencontré une situation qu'il ne sait pas traiter.",
      },
      "NotFound": {
        "content": {
          "application/json": {
            "schema": {
              "$ref": "#/components/schemas/Error",
            },
          },
        },
        "description": "Le serveur n'a pas trouvé la ressource demandée.",
      },
      "ServiceUnavailable": {
        "content": {
          "application/json": {
            "schema": {
              "$ref": "#/components/schemas/Error",
            },
          },
        },
        "description": "Le serveur n'est pas prêt pour traiter la requête",
      },
      "TooManyRequests": {
        "content": {
          "application/json": {
            "schema": {
              "$ref": "#/components/schemas/Error",
            },
          },
        },
        "description": "L'utilisateur a émis trop de requêtes dans un laps de temps donné.",
      },
      "Unauthorized": {
        "content": {
          "application/json": {
            "schema": {
              "$ref": "#/components/schemas/Error",
            },
          },
        },
        "description": "L'authentification est requise.",
      },
    },
    "schemas": {
      "Error": {
        "additionalProperties": false,
        "properties": {
          "code": {
            "type": [
              "string",
              "null",
            ],
          },
          "data": {
            "description": "Données contextuelles liées à l'erreur",
            "example": {
              "validationError": {
                "code": "FST_ERR_VALIDATION",
                "issues": [
                  {
                    "code": "invalid_type",
                    "expected": "number",
                    "message": "Number attendu",
                    "path": [
                      "longitude",
                    ],
                    "received": "nan",
                  },
                ],
                "name": "ZodError",
                "statusCode": 400,
                "validationContext": "querystring",
              },
            },
          },
          "message": {
            "description": "Un message explicatif de l'erreur",
            "example": "querystring.longitude: Number attendu",
            "type": "string",
          },
          "name": {
            "description": "Le type générique de l'erreur",
            "example": "Bad Request",
            "type": "string",
          },
          "statusCode": {
            "description": "Le status code retourné",
            "example": 400,
            "type": "number",
          },
        },
        "required": [
          "message",
          "name",
          "statusCode",
        ],
        "type": "object",
      },
    },
    "securitySchemes": {
      "api-key": {
        "in": "header",
        "name": "authorization",
        "type": "apiKey",
      },
    },
  },
  "info": {
    "license": {
      "name": "MIT",
    },
    "title": "API documentation BAL",
    "version": "V1.0",
  },
  "openapi": "3.1.0",
  "paths": {
    "/healthcheck": {
      "get": {
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "anyOf": [
                    {
                      "additionalProperties": false,
                      "description": "Organisation deca Response body",
                      "properties": {
                        "contrats": {
                          "additionalProperties": false,
                          "properties": {
                            "appr": {
                              "type": "number",
                            },
                            "prof": {
                              "type": "number",
                            },
                            "total": {
                              "type": "number",
                            },
                          },
                          "required": [
                            "total",
                            "appr",
                            "prof",
                          ],
                          "type": "object",
                        },
                        "dernier_contrat": {
                          "additionalProperties": false,
                          "properties": {
                            "date_debut_contrat": {
                              "type": "string",
                            },
                            "date_fin_contrat": {
                              "type": "string",
                            },
                          },
                          "type": [
                            "object",
                            "null",
                          ],
                        },
                        "premier_contrat": {
                          "additionalProperties": false,
                          "properties": {
                            "date_debut_contrat": {
                              "type": "string",
                            },
                            "date_fin_contrat": {
                              "type": "string",
                            },
                          },
                          "type": [
                            "object",
                            "null",
                          ],
                        },
                      },
                      "required": [
                        "contrats",
                      ],
                      "type": "object",
                    },
                    {
                      "additionalProperties": false,
                      "properties": {
                        "code": {
                          "type": [
                            "string",
                            "null",
                          ],
                        },
                        "data": {
                          "description": "Données contextuelles liées à l'erreur",
                          "example": {
                            "validationError": {
                              "code": "FST_ERR_VALIDATION",
                              "issues": [
                                {
                                  "code": "invalid_type",
                                  "expected": "number",
                                  "message": "Number attendu",
                                  "path": [
                                    "longitude",
                                  ],
                                  "received": "nan",
                                },
                              ],
                              "name": "ZodError",
                              "statusCode": 400,
                              "validationContext": "querystring",
                            },
                          },
                        },
                        "message": {
                          "description": "Un message explicatif de l'erreur",
                          "example": "querystring.longitude: Number attendu",
                          "type": "string",
                        },
                        "name": {
                          "description": "Le type générique de l'erreur",
                          "example": "Bad Request",
                          "type": "string",
                        },
                        "statusCode": {
                          "description": "Le status code retourné",
                          "example": 400,
                          "type": "number",
                        },
                      },
                      "required": [
                        "message",
                        "name",
                        "statusCode",
                      ],
                      "type": "object",
                    },
                  ],
                  "description": "Organisation deca Response body",
                },
              },
            },
            "description": "Organisation deca Response body",
          },
          "400": {
            "$ref": "#/components/responses/BadRequest",
          },
          "401": {
            "$ref": "#/components/responses/Unauthorized",
          },
          "403": {
            "$ref": "#/components/responses/Forbidden",
          },
          "404": {
            "$ref": "#/components/responses/NotFound",
          },
          "429": {
            "$ref": "#/components/responses/TooManyRequests",
          },
          "500": {
            "$ref": "#/components/responses/InternalServerError",
          },
          "503": {
            "$ref": "#/components/responses/ServiceUnavailable",
          },
        },
        "security": [],
        "tags": [
          "système",
        ],
      },
    },
    "/v1/deca/search/organisme": {
      "post": {
        "parameters": [
          {
            "description": "Bearer token",
            "in": "header",
            "name": "Authorization",
            "required": false,
            "schema": {
              "description": "Bearer token",
              "type": "string",
            },
          },
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "additionalProperties": false,
                "description": "Organisation deca Request body",
                "properties": {
                  "siret": {
                    "description": "Le numéro de SIRET de l'établissement",
                    "example": "78424186100011",
                    "pattern": "^[0-9]{14}$",
                    "type": "string",
                  },
                },
                "required": [
                  "siret",
                ],
                "type": "object",
              },
            },
          },
          "required": true,
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "anyOf": [
                    {
                      "additionalProperties": false,
                      "description": "Organisation deca Response body",
                      "properties": {
                        "contrats": {
                          "additionalProperties": false,
                          "properties": {
                            "appr": {
                              "type": "number",
                            },
                            "prof": {
                              "type": "number",
                            },
                            "total": {
                              "type": "number",
                            },
                          },
                          "required": [
                            "total",
                            "appr",
                            "prof",
                          ],
                          "type": "object",
                        },
                        "dernier_contrat": {
                          "additionalProperties": false,
                          "properties": {
                            "date_debut_contrat": {
                              "type": "string",
                            },
                            "date_fin_contrat": {
                              "type": "string",
                            },
                          },
                          "type": [
                            "object",
                            "null",
                          ],
                        },
                        "premier_contrat": {
                          "additionalProperties": false,
                          "properties": {
                            "date_debut_contrat": {
                              "type": "string",
                            },
                            "date_fin_contrat": {
                              "type": "string",
                            },
                          },
                          "type": [
                            "object",
                            "null",
                          ],
                        },
                      },
                      "required": [
                        "contrats",
                      ],
                      "type": "object",
                    },
                    {
                      "additionalProperties": false,
                      "properties": {
                        "code": {
                          "type": [
                            "string",
                            "null",
                          ],
                        },
                        "data": {
                          "description": "Données contextuelles liées à l'erreur",
                          "example": {
                            "validationError": {
                              "code": "FST_ERR_VALIDATION",
                              "issues": [
                                {
                                  "code": "invalid_type",
                                  "expected": "number",
                                  "message": "Number attendu",
                                  "path": [
                                    "longitude",
                                  ],
                                  "received": "nan",
                                },
                              ],
                              "name": "ZodError",
                              "statusCode": 400,
                              "validationContext": "querystring",
                            },
                          },
                        },
                        "message": {
                          "description": "Un message explicatif de l'erreur",
                          "example": "querystring.longitude: Number attendu",
                          "type": "string",
                        },
                        "name": {
                          "description": "Le type générique de l'erreur",
                          "example": "Bad Request",
                          "type": "string",
                        },
                        "statusCode": {
                          "description": "Le status code retourné",
                          "example": 400,
                          "type": "number",
                        },
                      },
                      "required": [
                        "message",
                        "name",
                        "statusCode",
                      ],
                      "type": "object",
                    },
                  ],
                  "description": "Organisation deca Response body",
                },
              },
            },
            "description": "Organisation deca Response body",
          },
          "400": {
            "$ref": "#/components/responses/BadRequest",
          },
          "401": {
            "$ref": "#/components/responses/Unauthorized",
          },
          "403": {
            "$ref": "#/components/responses/Forbidden",
          },
          "404": {
            "$ref": "#/components/responses/NotFound",
          },
          "429": {
            "$ref": "#/components/responses/TooManyRequests",
          },
          "500": {
            "$ref": "#/components/responses/InternalServerError",
          },
          "503": {
            "$ref": "#/components/responses/ServiceUnavailable",
          },
        },
        "security": [
          {
            "api-key": [],
          },
        ],
        "tags": [
          "v1",
        ],
      },
    },
    "/v1/organisation/validation": {
      "post": {
        "parameters": [
          {
            "description": "Bearer token",
            "in": "header",
            "name": "Authorization",
            "required": false,
            "schema": {
              "description": "Bearer token",
              "type": "string",
            },
          },
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "additionalProperties": false,
                "description": "Organisation validation Request body",
                "properties": {
                  "email": {
                    "format": "email",
                    "type": "string",
                  },
                  "siret": {
                    "description": "Le numéro de SIRET de l'établissement",
                    "example": "78424186100011",
                    "pattern": "^[0-9]{14}$",
                    "type": "string",
                  },
                },
                "required": [
                  "email",
                  "siret",
                ],
                "type": "object",
              },
            },
          },
          "required": true,
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "anyOf": [
                    {
                      "additionalProperties": false,
                      "description": "Organisation deca Response body",
                      "properties": {
                        "contrats": {
                          "additionalProperties": false,
                          "properties": {
                            "appr": {
                              "type": "number",
                            },
                            "prof": {
                              "type": "number",
                            },
                            "total": {
                              "type": "number",
                            },
                          },
                          "required": [
                            "total",
                            "appr",
                            "prof",
                          ],
                          "type": "object",
                        },
                        "dernier_contrat": {
                          "additionalProperties": false,
                          "properties": {
                            "date_debut_contrat": {
                              "type": "string",
                            },
                            "date_fin_contrat": {
                              "type": "string",
                            },
                          },
                          "type": [
                            "object",
                            "null",
                          ],
                        },
                        "premier_contrat": {
                          "additionalProperties": false,
                          "properties": {
                            "date_debut_contrat": {
                              "type": "string",
                            },
                            "date_fin_contrat": {
                              "type": "string",
                            },
                          },
                          "type": [
                            "object",
                            "null",
                          ],
                        },
                      },
                      "required": [
                        "contrats",
                      ],
                      "type": "object",
                    },
                    {
                      "additionalProperties": false,
                      "properties": {
                        "code": {
                          "type": [
                            "string",
                            "null",
                          ],
                        },
                        "data": {
                          "description": "Données contextuelles liées à l'erreur",
                          "example": {
                            "validationError": {
                              "code": "FST_ERR_VALIDATION",
                              "issues": [
                                {
                                  "code": "invalid_type",
                                  "expected": "number",
                                  "message": "Number attendu",
                                  "path": [
                                    "longitude",
                                  ],
                                  "received": "nan",
                                },
                              ],
                              "name": "ZodError",
                              "statusCode": 400,
                              "validationContext": "querystring",
                            },
                          },
                        },
                        "message": {
                          "description": "Un message explicatif de l'erreur",
                          "example": "querystring.longitude: Number attendu",
                          "type": "string",
                        },
                        "name": {
                          "description": "Le type générique de l'erreur",
                          "example": "Bad Request",
                          "type": "string",
                        },
                        "statusCode": {
                          "description": "Le status code retourné",
                          "example": 400,
                          "type": "number",
                        },
                      },
                      "required": [
                        "message",
                        "name",
                        "statusCode",
                      ],
                      "type": "object",
                    },
                  ],
                  "description": "Organisation deca Response body",
                },
              },
            },
            "description": "Organisation deca Response body",
          },
          "400": {
            "$ref": "#/components/responses/BadRequest",
          },
          "401": {
            "$ref": "#/components/responses/Unauthorized",
          },
          "403": {
            "$ref": "#/components/responses/Forbidden",
          },
          "404": {
            "$ref": "#/components/responses/NotFound",
          },
          "429": {
            "$ref": "#/components/responses/TooManyRequests",
          },
          "500": {
            "$ref": "#/components/responses/InternalServerError",
          },
          "503": {
            "$ref": "#/components/responses/ServiceUnavailable",
          },
        },
        "security": [
          {
            "api-key": [],
          },
        ],
        "tags": [
          "v1",
        ],
      },
    },
  },
  "servers": [
    {
      "description": "Production",
      "url": "https://bal.apprentissage.beta.gouv.fr",
    },
  ],
  "webhooks": {},
}
`;
